### PR TITLE
ConcurrentSubscripiton avoid concurrent access for invalid demand

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
ConcurrentSubscription currently propagates invalid demand without any
concurrency protection. In general this is invalid use of the API but
may invalidate underlying data structures that are not thread safe and
result in undefined results.

Modifications:
- Use a simpler locking scheme inspired by Publisher#flatMapMerge design
which allows for re-entry and also notification is another thread has
attempted to acquire the lock which will trigger re-processing.

Result:
ConcurrentSubscripiton no longer allows any concurrent access and uses a
more common/shareable locking utility.

### CODE REVIEW

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java [hunk 1]
```diff
@@ -36,6 +37,49 @@ private ConcurrentUtils() {
...
+     * {@link #acquirePendingReentrantLock(AtomicLongFieldUpdater, Object)}.
+     * @param owner The owner of the lock object.
+     * @param <T> The type of object that owns the lock.
+     * @return {@code true} if the lock was released, or this method call corresponds to a prior re-entrant call
+     * to {@link #acquirePendingReentrantLock(AtomicLongFieldUpdater, Object)}.
+     */
+    public static <T> boolean releasePendingReentrantLock(final AtomicLongFieldUpdater<T> lockUpdater,
+                                                          final long acquireId, final T owner) {
+        assert acquireId != 0;
+        return acquireId < 0 || lockUpdater.getAndSet(owner, 0) == acquireId;
```
Comment (normanmaurer) [inline servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java:80]: nit: consider define a static field for `0` so its a bit less of a magic number

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java [hunk 2]
```diff
@@ -71,74 +74,59 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
...
+        if (acquirePendingReentrantLock(subscriptionLockUpdater, this) != 0) {
             subscription.cancel();
+            // poison subscriptionLockUpdater
         }
     }
+
+    private static long mapInvalidRequestN(long n) {
+        // We map zero to a negative number because zero could later be overwritten by a subsequent legit value of
+        // n, and we want to ensure the invalid use gets propagated.
+        return n == CANCELLED ? CANCELLED + 1 : n == 0 ? -1 : n;
```
Comment (normanmaurer) [inline servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:130]: nit: Imho chaining stuff like this is really not readable. Consider using a branch

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java [hunk 3]
```diff
@@ -71,74 +74,59 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
...
+                final long prevPendingDemand = pendingDemandUpdater.getAndSet(this, 0);
+                if (prevPendingDemand == CANCELLED) {
+                    subscription.cancel();
+                } else if (prevPendingDemand != 0) {
+                    subscription.request(prevPendingDemand);
+                }
+            } catch (Throwable cause) {
+                if (delayedCause == null) {
+                    delayedCause = cause;
+                }
```
Comment (normanmaurer) [inline servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:105]: should we use `addSuppressed(...)` if `delayCause != null` ?

File: servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapSingleTest.java [hunk 4]
```diff
@@ -67,7 +67,7 @@
 
 public class PublisherFlatMapSingleTest {
     @Rule
-    public final Timeout timeout = new ServiceTalkTestTimeout(30, SECONDS);
+    public final Timeout timeout = new ServiceTalkTestTimeout();
```
Comment (normanmaurer) [inline servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapSingleTest.java:70]: nit: why this change ?

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java (hunk 1 above)
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java]: done

File: servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapSingleTest.java (hunk 4 above)
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapSingleTest.java:70]: the test doesn't need a special timeout and can use the default value. we now have a higher default value for CI that didn't exist when this timeout was originally configured.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 3 above)
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:105]: lets discuss on https://github.com/apple/servicetalk/pull/1011#discussion_r408440046

**Comment by Scottmitch** [comment]
@servicetalk-bot - test this please

**Comment by Scottmitch** [comment]
humm ... ci failures
```
Creating servicetalk-java8-prb_runtime-setup_1 ... done
Error: No such container: f7df93647ccd8b7580a740e917771dbc1589b9f79d8694758bcf73a244e776d8
21:45:17 No such container: f7df93647ccd8b7580a740e917771dbc1589b9f79d8694758bcf73a244e776d8
21:45:17 Build step 'Execute shell' marked build as failure
```

**Comment by Scottmitch** [comment]
@servicetalk-bot - test this please

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 2 above)
Comment (normanmaurer) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:125]: @Scottmitch ^^

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/744

**Comment by Scottmitch** [comment]
@servicetalk-bot - test this please

**Comment by Scottmitch** [comment]
ci failure...
```
08:28:08 
Creating servicetalk-java8-prb_runtime-setup_1 ... done
time="2020-04-28T15:28:09Z" level=error msg="error waiting for container: context canceled"
08:28:09 Error response from daemon: failed to get network during CreateEndpoint: network 82cd76a80b00834668362278f1efff2e358505fff034522ffee45181c158dfe9 not found
08:28:09 Build step 'Execute shell' marked build as failure
```

**Comment by Scottmitch** [comment]
@servicetalk-bot - test this please

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java [hunk 5]
```diff
@@ -65,6 +76,49 @@ private ConcurrentUtils() {
...
+     */
+    public static <T> long tryAcquireReentrantLock(final AtomicLongFieldUpdater<T> lockUpdater, final T owner) {
+        final long threadId = Thread.currentThread().getId();
+        for (;;) {
+            final long prevThreadId = lockUpdater.get(owner);
+            if (prevThreadId == REENTRANT_LOCK_ZERO_THREAD_ID) {
+                if (lockUpdater.compareAndSet(owner, REENTRANT_LOCK_ZERO_THREAD_ID, threadId)) {
+                    return threadId;
+                }
+            } else if (prevThreadId == threadId || prevThreadId == -threadId) {
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java:96]: A thread which acquires the lock cannot be "terminated" (e.g. complete execution) without releasing the lock as the thread is currently executing the code inside the `try {} finally {}` scope where the lock is held. So either this is the same thread in a reentrant situation, or it is a different thread which will have a different ID (as thread IDs are unique for live threads).

`Thread#stop` exists which in theory could called by a method while the lock is held, but the thread doesn't "terminate" until exiting the stack. Also `Thread#stop` is `@Deprecated` and marked as `inherently unsafe ... potentially resulting in arbitrary behavior`.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java [hunk 6]
```diff
@@ -71,74 +70,58 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
...
-                        return;
-                    } else {
-                        continue;
-                    }
+            pendingDemand = mapInvalidRequestN(n);
+        } else {
+            for (;;) {
+                final long prevPendingDemand = pendingDemand;
+                if (prevPendingDemand < 0 ||
+                        pendingDemandUpdater.compareAndSet(this, prevPendingDemand,
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:78]: hum good point this should be the equivalent. only difference is with a negative value here we just move on (no CAS), `accumulateAndGet` will still do the CAS-loop, but this is fine.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java [hunk 7]
```diff
@@ -71,74 +70,58 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
...
             try {
-                subscription.request(n);
+                final long prevPendingDemand = pendingDemandUpdater.getAndSet(this, 0);
+                if (prevPendingDemand == CANCELLED) {
+                    subscription.cancel();
+                } else if (prevPendingDemand != 0) {
+                    subscription.request(prevPendingDemand);
+                }
+            } catch (Throwable cause) {
+                if (delayedCause == null) {
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:100]: A general characteristics of the ConcurrentUtil locks is if an exception occurs while we hold the lock, we have to successfully release the lock otherwise we may miss events from other threads. In this case that mean we miss a `cancel` which is not desirable. We could assume it is the caller's responsibility to recover/cancel but if the cleanup has already been done from the caller's perspective it may not happen again.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java [hunk 8]
```diff
@@ -71,74 +70,58 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
...
+                    delayedCause = cause;
+                }
             } finally {
-                subscriptionLockOwner = null;
+                tryAcquire = !releaseReentrantLock(subscriptionLockUpdater, acquireId, this);
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:104]: lets discuss above. I don't think we want todo this.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java [hunk 9]
```diff
@@ -71,74 +70,58 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
...
                 }
             }
-            if (n < 0) {
-                subscription.cancel();
-                return; // Don't set subscriptionLockOwner = 0 ... we don't want to request any more!
+        }
+        Throwable delayedCause = null;
+        boolean tryAcquire;
+        do {
+            final long acquireId = tryAcquireReentrantLock(subscriptionLockUpdater, this);
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:87]: good point, I also considered this. the re-entrant lock isn't strictly necessary for correctness as we could use the non-reentrant lock which also requires re-acquire semantics. for re-entry scenarios this approach saves a CAS on each re-entrant operation and trades it for a `Thread.getId()` (which is thread local state). Current approach preserves the previous re-entrant semantics in this class to minimize behavior difference and also shows perf improvement:

```
# Reentrant lock (this PR)
Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  47753164.225 ± 850903.330  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    714751.803 ±   8155.984  ops/s

Benchmark                                     Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  48522938.411 ± 1065238.470  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    723176.975 ±   11730.345  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  48460738.326 ± 918765.253  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    718663.755 ±  22510.103  ops/s

# Non-reentrant (this PR modified to use tryAcquireLock)
Benchmark                                     Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  46236169.698 ± 1304606.373  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    379186.656 ±    7327.488  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  46554069.756 ± 770443.754  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    374243.264 ±   4981.803  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  46420860.395 ± 906316.830  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    379715.245 ±   7939.865  ops/s
```

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 2 above)
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:116]: I prefer to keep the code as is as it is more compact and I don't think readability is dramatically different between the two options:

current option:
```
return n == CANCELLED ? CANCELLED + 1 : n == 0 ? -1 : n;
```

suggested option:
```
if (n == CANCELLED) {
  return CANCELLED + 1;
}
return n == 0 ? -1 : n;
```

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java]: Practically, we are discussing the case when `subscription.request()` or `subscription.cancel()` throws which in turn makes this `request()` method throws. 

Once a `Subscription` throws thus violating the spec, any expectation around further interaction with the `Subscription` doing anything meaningful isn't really justified. If this helps us simplify control-flow, I think it is a valid approach to take.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:80]: > Current approach preserves the previous re-entrant semantics in this class to minimize behavior difference

We are modifying an internal class in a way that adheres to the specifications. If we do not depend on this behavior (which we shouldn't), I don't think there is anything wrong in simplification. Allowing re-entrancy here adds more complexity in this code and additionally in `ConcurrentUtils`. Removing re-entrancy reduces thread-stack which is a net win. 

I don't think the small perf win trumps the other benefits.

[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:109]: The benchmark was helpful :) detected regression from master on the no-reentrant case. master skips the atomic long accumulator if we are able to acquire the lock, but trades some concurrency. I added a fast path when no concurrency is detected and we now gain in both cases (re-entrant and non-reentrant) which is assumed to be more common, and even if there is concurrency the expected cost is higher.

```
# Re-entrant with fast path
Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  74805386.276 ± 746665.173  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5   1895818.132 ±  31661.798  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  74602546.497 ± 972834.986  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5   2035494.125 ±  28986.077  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  74724470.463 ± 809942.225  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5   1864583.823 ±  48146.217  ops/s

# master
Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  59009939.298 ± 913522.147  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    449874.916 ±  10856.180  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  58815916.826 ± 963463.757  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    447968.057 ±   6171.099  ops/s

Benchmark                                     Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  58929994.344 ± 1208720.403  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    446883.247 ±   15198.500  ops/s
```

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java]: > any expectation around further interaction with the Subscription doing anything meaningful isn't really justified.

Agreed in general the RS spec is violated but this may happen in practice and will be challenging to debug if a pending/subsequent `cancel()` is not propagated because an in process `request(..)` throws. The concurrency is introduced by this class and can be encapsulated in the class and IMHO the simplification on control flow is not worth the potential complexity to debug later. The control. flow is also different between `request` and `cancel` so it isn't necessarily the case that an exception/bug in `request` doesn't mean `cancel` also has a bug. Making a best effort to propagate means the problem would be debugged elsewhere and won't require unrolling concurrent access trying to understand when/how signals were dropped.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:109]: Here are the numbers for the `tryAcquireLock` version with fast path. The noReentrant is similar as expected now but the re-entrant case is much worse (only ~18% of the perf). In order to avoid regression I think it makes sense to move forward with the approach in this PR (as opposed to using `tryAcquireLock`).

```
# Non-reentrant (this PR with fast path, modified to use tryAcquireLock)
Benchmark                                     Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  75341058.045 ± 1011207.561  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    337948.564 ±    5698.630  ops/s
```

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java]: I agree that the control-flow simplification isn't substantial. 

You are correct that the following sequence of events is safe to allow:

```
request(1) <- throws
cancel() <- passes
```

but are the following sequence of events also safe or improve debuggability:

```
request(1) <- throws
request(1) <- passes

now, the Subscriber is expecting 2 items but only will ever get 1
```

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:109]: From my perspective, breaking recursion here is better than allowing it as it cuts down the thread stack. 

I would suggest avoiding the complexity of adding reentrant lock semantics in `ConcurrentUtils`. I am also not completely convinced about the approach for the reentrant lock but we can discuss that in https://github.com/apple/servicetalk/pull/1015#discussion_r417523512

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java]: the second sequence is in the undefined behavior territory. I think reasonable behavior for `ConcurrentSubscription`, which is an intermediate layer, is to deliver signals as it receives them instead of dropping/filtering signals because it doesn't have scope to know if they will be useful or not (e.g. debugging, making progress, etc.). The canonical usage of the `ConcurrentUtil` locks facilitate this (e.g. `while(tryAcquire) { try { /* critical section */ } finally { release } }`) without custom exception handling. Are you suggesting actively preventing scenario 2 but allowing scenario 1, and what would we gain (we are in unexpected error territory)?

**Comment by Scottmitch** [comment]
suspected CI related failure ... https://ci.servicetalk.io/job/servicetalk-java11-prb/1280/consoleText
```
BUILD SUCCESSFUL in 9m 56s
301 actionable tasks: 242 executed, 59 up-to-date
removal of container ad39f54ec9fb47d45e167e146414ef1ca48efa7f3bc34ca711a307b6f548bfa3 is already in progress
Build step 'Execute shell' marked build as failure
```

**Comment by Scottmitch** [comment]
@servicetalk-bot - test this please

**Comment by Scottmitch** [comment]
Benchmarks were revised to provide multi-threaded perf for non-reentrant case and more stack depth options for recursion.

Summary:
#### `tryAcquireReentrantLock`
- gains against `master` for reentrant (+151% to +500% depending upon recursion depth)
- gains against `tryAcquireLock` for reentrant (bigger margins than master)
- gains against `master` for single threaded non-reentrant (+28%)
- similar to `master` for multi threaded non-reentrant
- similar to `tryAcquireLock ` for non-reentrant
#### `tryAcquireLock`
- regresses against `master` for reentrant (-10% to -30% depending upon recursion depth)

```
## master
Benchmark                                              (stackDepth)   Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionReentrantBenchmark.singleThread             2  thrpt    5  24318761.311 ± 965534.969  ops/s
ConcurrentSubscriptionReentrantBenchmark.singleThread            10  thrpt    5   4608999.521 ± 475365.160  ops/s
ConcurrentSubscriptionReentrantBenchmark.singleThread           100  thrpt    5    453085.063 ±  53186.602  ops/s

Benchmark                                      Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.multiThread   thrpt    5   6753766.730 ± 477061.905  ops/s
ConcurrentSubscriptionBenchmark.singleThread  thrpt    5  58855485.836 ± 715964.377  ops/s

## Reentrant (using tryAcquireReentrantLock)
Benchmark                                              (stackDepth)   Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionReentrantBenchmark.singleThread             2  thrpt    5  61086581.693 ± 866190.998  ops/s
ConcurrentSubscriptionReentrantBenchmark.singleThread            10  thrpt    5  27976426.295 ± 842739.171  ops/s
ConcurrentSubscriptionReentrantBenchmark.singleThread           100  thrpt    5   2248121.486 ±  54539.770  ops/s

Benchmark                                      Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.multiThread   thrpt    5   6491110.907 ±  317372.027  ops/s
ConcurrentSubscriptionBenchmark.singleThread  thrpt    5  76085499.345 ± 1123039.174  ops/s

## Non-reentrant (using tryAcquireLock)
Benchmark                                              (stackDepth)   Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionReentrantBenchmark.singleThread             2  thrpt    5  21694566.056 ± 2973401.571  ops/s
ConcurrentSubscriptionReentrantBenchmark.singleThread            10  thrpt    5   3410420.825 ±  384962.783  ops/s
ConcurrentSubscriptionReentrantBenchmark.singleThread           100  thrpt    5    323562.643 ±   37266.246  ops/s

Benchmark                                      Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.multiThread   thrpt    5   6423213.603 ±  601263.682  ops/s
ConcurrentSubscriptionBenchmark.singleThread  thrpt    5  75500536.822 ± 5127380.129  ops/s
```

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:109]: It isn't the responsibility of this class to limit recursion/stack depth. Limiting stack depth is shown in the benchmarks to reduce performance.

File: servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/ConcurrentUtilsTest.java [hunk 10]
```diff
@@ -56,13 +66,22 @@ public void tearDown() throws Exception {
...
     }
 
     @Test
-    public void pendingFromDifferentThread() throws Exception {
+    public void reentrantLockSingleThread() {
+        long acquireId = tryAcquireReentrantLock(reentrantLockUpdater, this);
+        assertThat(acquireId, greaterThan(0L));
+        long acquireId2 = tryAcquireReentrantLock(reentrantLockUpdater, this);
+        assertThat(acquireId2, is(-acquireId));
+        assertTrue(releaseReentrantLock(reentrantLockUpdater, acquireId, this));
```
Comment (NiteshKant) [inline servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/ConcurrentUtilsTest.java:80]: Add an assert for releasing `acquiredId2`?

`assertTrue(releaseReentrantLock(reentrantLockUpdater, acquireId2, this));`

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java:96]: Ok so I think this means that this method must only be called when used with `try-finally` and not across thread boundaries like we do in `NettyPipelinedConnection`.  If there are inherent assumptions like this then either we should mention it in the method name or in javadocs.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:123]: I am still not sure about the approach here of continuing to call the `subscription` if a previous call to it threw. If we are not comfortable about poisoning the lock then an alternate is to set the `pending = CANCELLED` and hence ignoring any subsequent calls. If we do this, then the fast-path in `request(n)` can also avoid calling `drain()` from inside a `finally` block.

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/1040

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java:96]: Good point, there are some additional constraints worth clarifying.

> This lock <strong>must</strong> eventually be released by the same thread that acquired the lock. If the thread
 that acquires this lock is terminated before releasing the lock state is undefined.

fwiw the lock can be used across thread boundaries, and shared characteristics with the non-reentrant lock where usage outside of a try/finally can lead to undefined state (e.g. missed unlock), however the interesting part here is if the thread which acquires the lock is terminated before release that leads to an undefined state (clarified in comments above).

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java:123]: > If we do this, then the fast-path in request(n) can also avoid calling drain() from inside a finally block.

We have two options for reacting to exceptions:
1. force cleanup internally to `ConcurrentSubscription`
2. allow cleanup to happen externally to `ConcurrentSubscription`

for (1) we have to catch the exception, force a call to `cancel()`, and we can optionally release the lock. This would require a `catch` in the fast path `request(n)` method, and a catch in the `drain()`, and we could release the lock outside the `finally`.

for (2) we have to catch/ignore exceptions while we hold the lock (because `cancel()` may come concurrently). this is the general pattern of the `ConcurrentUtil` locks in which you keep draining until you can release the lock, otherwise you may miss an event.

I would like to avoid forcing `cancel()` from `ConcurrentSubscription` as it is an interim layer and not the root cause of the exception. If the `ConcurrentSubscription` was removed from the scenario the caller would still have the same concerns related to dealing with exceptions.

**Comment by Scottmitch** [comment]
discussed relaxing the exception handling with @NiteshKant offline. I've modified this PR to incorporate @NiteshKant's suggestions and we can enhance later if necessary.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java]: discussed offline https://github.com/apple/servicetalk/pull/1015#issuecomment-623791000
